tools/firtool/firtool.cpp

```diff
@@ -381,6 +382,9 @@ static LogicalResult processBuffer(
     if (exportModuleHierarchy)
       pm.addPass(sv::createHWExportModuleHierarchyPass(outputFilename));
 
+    // Check inner symbols and inner refs.
+    pm.addPass(hw::createVerifyInnerRefNamespacePass());
```

Comment:
The idea here is that this will catch some errors that could manifest during `ExportVerilog`? This may miss some issues. It is better than nothing.

I'm lukewarm on this approach as it reminds me of SFC-style verification. We used to do all the verification in a [block of serial passes](https://github.com/chipsalliance/firrtl/blob/1.6.x/src/main/scala/firrtl/stage/Forms.scala#L33) that was run once and never again. I ran experiments where these passes would be rescheduled after every pass (normal "verification") and it uncovered a lot of really weird issues.

tl;dr: I'm okay with this, but only as an intermediate step. Don't let this hang around for too long. 😉 We need to bite the bullet and get the container op in place.

Reply:
+1 on the container op

Reply:
100% I'd prefer the container op.  However we're increasingly using this and getting some verification in place seems high-priority to avoid things being too broken without noticing.

> The idea here is that this will catch some errors that could manifest during ExportVerilog? This may miss some issues. It is better than nothing.

Yes, idea was to schedule it in a few places to help catch things -- right now this PR adds it as part of the FIRRTL->HW pipeline, the HW->SV pipeline, and yep a final run before `ExportVerilog`.

Probably could automagically insert this between all passes or something (drive via pass instrumentation?) if the concern is only having a one-off check being performed.  Strong agree I'd prefer to verify all the things regularly :+1:.

As an aside there's a lot of "walk this module looking for the symbol" logic that can be improved once all this in place.

> tl;dr: I'm okay with this, but only as an intermediate step. Don't let this hang around for too long. 😉 We need to bite the bullet and get the container op in place.

Absolutely and thanks for saying so! The op itself is easy, getting everything to use and maintain that sense of structure vs bag of operations is a bit more challenging.  Haven't pushed on it for a few days due to other demands, but will revisit and see if can't form a migration plan. :+1: .